File: app/src/test/java/com/lifetrace/execute/data/repository/ProjectRepositoryTest.kt
package com.lifetrace.execute.data.repository

import android.app.Application
import android.content.Context
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import com.lifetrace.execute.data.local.LifeTraceExecuteDatabase
import com.lifetrace.execute.domain.project.ExecutionProjectStatus
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [35], application = Application::class)
class ProjectRepositoryTest {
    private lateinit var database: LifeTraceExecuteDatabase
    private lateinit var projects: ProjectRepository
    private lateinit var tasks: TaskRepository

    @Before
    fun setUp() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        database = Room.inMemoryDatabaseBuilder(context, LifeTraceExecuteDatabase::class.java)
            .allowMainThreadQueries()
            .build()
        projects = ProjectRepository(database)
        tasks = TaskRepository(database)
    }

    @After
    fun tearDown() = database.close()

    @Test
    fun createAndUpdateAreLocalFirstAndQueueOutbox() = runBlocking {
        val project = projects.createProject(
            userId = "user-1",
            deviceId = "device-a",
            name = "  Execute 1.0  ",
            description = "  ship full product  ",
        )
        assertEquals("Execute 1.0", project.name)
        assertNotNull(database.dao().getProject("user-1", project.id))
        assertEquals(1, database.dao().observePendingOutboxCount("user-1").first())

        val updated = projects.updateProject(
            project = project,
            deviceId = "device-b",
            status = ExecutionProjectStatus.PAUSED,
        )
        assertEquals(2, updated.localVersion)
        assertEquals("paused", database.dao().getProject("user-1", project.id)!!.status)
        assertEquals(2, database.dao().observePendingOutboxCount("user-1").first())
    }

    @Test
    fun deletingProjectWithLinkedTaskIsRejected() = runBlocking {
        val project = projects.createProject("user-1", "device-a", "Project")
        tasks.createTask(
            userId = "user-1",
            deviceId = "device-a",
            title = "Linked task",
            projectId = project.id,
        )

        assertThrows(IllegalArgumentException::class.java) {
            runBlocking { projects.deleteProject("user-1", project.id) }
        }
        assertNotNull(database.dao().getProject("user-1", project.id))
    }

    @Test
    fun unreferencedProjectDeleteCreatesTombstoneIntent() = runBlocking {
        val project = projects.createProject("user-1", "device-a", "Disposable")
        projects.deleteProject("user-1", project.id)

        assertNull(database.dao().getProject("user-1", project.id))
        assertEquals(2, database.dao().observePendingOutboxCount("user-1").first())
    }
}

File: flutter_app/test/project_repository_test.dart
import 'package:drift/drift.dart' hide isNull;
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:lifetrace_execute/data/local/app_database.dart';
import 'package:lifetrace_execute/data/repository/project_repository.dart';
import 'package:lifetrace_execute/data/repository/task_repository.dart';
import 'package:lifetrace_execute/domain/project/execution_project.dart';

void main() {
  late AppDatabase database;
  late DriftProjectRepository projects;
  late DriftTaskRepository tasks;

  setUp(() {
    database = AppDatabase(NativeDatabase.memory());
    projects = DriftProjectRepository(database);
    tasks = DriftTaskRepository(database);
  });

  tearDown(() async => database.close());

  test('project CRUD persists locally and writes execution.project outbox', () async {
    final created = await projects.createProject(
      userId: 'user-1',
      deviceId: 'device-1',
      title: '  Real Project  ',
      description: 'local first',
      dueAt: '2026-10-01T00:00:00.000Z',
    );

    expect(created.title, 'Real Project');
    expect(await database.select(database.projects).get(), hasLength(1));
    var outbox = await database.select(database.syncOutbox).get();
    expect(outbox, hasLength(1));
    expect(outbox.single.entityType, DriftProjectRepository.entityType);

    final updated = await projects.updateProject(
      project: created,
      deviceId: 'device-1',
      status: ExecutionProjectStatus.paused,
    );
    expect(updated.localVersion, 2);
    expect(
      (await database.select(database.projects).get()).single.status,
      'paused',
    );

    await projects.deleteProject(project: updated, deviceId: 'device-1');
    expect(await database.select(database.projects).get(), isEmpty);
    outbox = await database.select(database.syncOutbox).get();
    expect(outbox.last.operation, 'delete');
    expect(outbox.last.entityType, DriftProjectRepository.entityType);
  });

  test('deleting a project unlinks its tasks and queues task update', () async {
    final project = await projects.createProject(
      userId: 'user-1',
      deviceId: 'device-1',
      title: 'Project with task',
    );
    final task = await tasks.createTask(
      userId: 'user-1',
      deviceId: 'device-1',
      title: 'Linked task',
      projectId: project.id,
    );

    await projects.deleteProject(project: project, deviceId: 'device-1');

    final storedTask = await (database.select(database.tasks)
          ..where((row) => row.id.equals(task.id)))
        .getSingle();
    expect(storedTask.projectId, isNull);

    final outbox = await database.select(database.syncOutbox).get();
    final taskChanges = outbox
        .where(
          (row) =>
              row.entityType == DriftTaskRepository.entityType &&
              row.entityId == task.id,
        )
        .toList();
    expect(taskChanges, hasLength(2));
    expect(taskChanges.last.payloadJson, contains('"projectId":null'));
  });
}
